diff --git a/packages/app-store/googlecalendar/lib/CalendarService.ts b/packages/app-store/googlecalendar/lib/CalendarService.ts
index 939d18e42..a6a0c2f39 100644
--- a/packages/app-store/googlecalendar/lib/CalendarService.ts
+++ b/packages/app-store/googlecalendar/lib/CalendarService.ts
@@ -18,6 +18,8 @@ import type {
 } from "@calcom/types/Calendar";
 import type { CredentialPayload } from "@calcom/types/Credential";
 
+import parseRefreshTokenResponse from "../../_utils/oauth/parseRefreshTokenResponse";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import { getGoogleAppKeys } from "./getGoogleAppKeys";
 import { googleCredentialSchema } from "./googleCredentialSchema";
 
@@ -81,11 +83,18 @@ export default class GoogleCalendarService implements Calendar {
 
     const refreshAccessToken = async (myGoogleAuth: Awaited<ReturnType<typeof getGoogleAuth>>) => {
       try {
-        const { res } = await myGoogleAuth.refreshToken(googleCredentials.refresh_token);
+        const res = await refreshOAuthTokens(
+          async () => {
+            const fetchTokens = await myGoogleAuth.refreshToken(googleCredentials.refresh_token);
+            return fetchTokens.res;
+          },
+          "google-calendar",
+          credential.userId
+        );
         const token = res?.data;
         googleCredentials.access_token = token.access_token;
         googleCredentials.expiry_date = token.expiry_date;
-        const key = googleCredentialSchema.parse(googleCredentials);
+        const key = parseRefreshTokenResponse(googleCredentials, googleCredentialSchema);
         await prisma.credential.update({
           where: { id: credential.id },
           data: { key },
diff --git a/packages/app-store/hubspot/lib/CalendarService.ts b/packages/app-store/hubspot/lib/CalendarService.ts
index b9817c616..66095d6de 100644
--- a/packages/app-store/hubspot/lib/CalendarService.ts
+++ b/packages/app-store/hubspot/lib/CalendarService.ts
@@ -23,6 +23,7 @@ import type {
 import type { CredentialPayload } from "@calcom/types/Credential";
 
 import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import type { HubspotToken } from "../api/callback";
 
 const hubspotClient = new hubspot.Client();
@@ -173,13 +174,18 @@ export default class HubspotCalendarService implements Calendar {
 
     const refreshAccessToken = async (refreshToken: string) => {
       try {
-        const hubspotRefreshToken: HubspotToken = await hubspotClient.oauth.tokensApi.createToken(
-          "refresh_token",
-          undefined,
-          WEBAPP_URL + "/api/integrations/hubspot/callback",
-          this.client_id,
-          this.client_secret,
-          refreshToken
+        const hubspotRefreshToken: HubspotToken = await refreshOAuthTokens(
+          async () =>
+            await hubspotClient.oauth.tokensApi.createToken(
+              "refresh_token",
+              undefined,
+              WEBAPP_URL + "/api/integrations/hubspot/callback",
+              this.client_id,
+              this.client_secret,
+              refreshToken
+            ),
+          "hubspot",
+          credential.userId
         );
 
         // set expiry date as offset from current time.
diff --git a/packages/app-store/larkcalendar/lib/CalendarService.ts b/packages/app-store/larkcalendar/lib/CalendarService.ts
index 88dba3fc2..6e3edaf0b 100644
--- a/packages/app-store/larkcalendar/lib/CalendarService.ts
+++ b/packages/app-store/larkcalendar/lib/CalendarService.ts
@@ -11,6 +11,7 @@ import type {
 } from "@calcom/types/Calendar";
 import type { CredentialPayload } from "@calcom/types/Credential";
 
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import { handleLarkError, isExpired, LARK_HOST } from "../common";
 import type {
   CreateAttendeesResp,
@@ -63,17 +64,22 @@ export default class LarkCalendarService implements Calendar {
     }
     try {
       const appAccessToken = await getAppAccessToken();
-      const resp = await fetch(`${this.url}/authen/v1/refresh_access_token`, {
-        method: "POST",
-        headers: {
-          Authorization: `Bearer ${appAccessToken}`,
-          "Content-Type": "application/json; charset=utf-8",
-        },
-        body: JSON.stringify({
-          grant_type: "refresh_token",
-          refresh_token: refreshToken,
-        }),
-      });
+      const resp = await refreshOAuthTokens(
+        async () =>
+          await fetch(`${this.url}/authen/v1/refresh_access_token`, {
+            method: "POST",
+            headers: {
+              Authorization: `Bearer ${appAccessToken}`,
+              "Content-Type": "application/json; charset=utf-8",
+            },
+            body: JSON.stringify({
+              grant_type: "refresh_token",
+              refresh_token: refreshToken,
+            }),
+          }),
+        "lark-calendar",
+        credential.userId
+      );
 
       const data = await handleLarkError<RefreshTokenResp>(resp, this.log);
       this.log.debug(
diff --git a/packages/app-store/office365calendar/lib/CalendarService.ts b/packages/app-store/office365calendar/lib/CalendarService.ts
index 39e6ceb98..5742fde8f 100644
--- a/packages/app-store/office365calendar/lib/CalendarService.ts
+++ b/packages/app-store/office365calendar/lib/CalendarService.ts
@@ -17,6 +17,8 @@ import type {
 } from "@calcom/types/Calendar";
 import type { CredentialPayload } from "@calcom/types/Credential";
 
+import parseRefreshTokenResponse from "../../_utils/oauth/parseRefreshTokenResponse";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import type { O365AuthCredentials } from "../types/Office365Calendar";
 import { getOfficeAppKeys } from "./getOfficeAppKeys";
 
@@ -241,28 +243,25 @@ export default class Office365CalendarService implements Calendar {
 
     const refreshAccessToken = async (o365AuthCredentials: O365AuthCredentials) => {
       const { client_id, client_secret } = await getOfficeAppKeys();
-      const response = await fetch("https://login.microsoftonline.com/common/oauth2/v2.0/token", {
-        method: "POST",
-        headers: { "Content-Type": "application/x-www-form-urlencoded" },
-        body: new URLSearchParams({
-          scope: "User.Read Calendars.Read Calendars.ReadWrite",
-          client_id,
-          refresh_token: o365AuthCredentials.refresh_token,
-          grant_type: "refresh_token",
-          client_secret,
-        }),
-      });
+      const response = await refreshOAuthTokens(
+        async () =>
+          await fetch("https://login.microsoftonline.com/common/oauth2/v2.0/token", {
+            method: "POST",
+            headers: { "Content-Type": "application/x-www-form-urlencoded" },
+            body: new URLSearchParams({
+              scope: "User.Read Calendars.Read Calendars.ReadWrite",
+              client_id,
+              refresh_token: o365AuthCredentials.refresh_token,
+              grant_type: "refresh_token",
+              client_secret,
+            }),
+          }),
+        "office365-calendar",
+        credential.userId
+      );
       const responseJson = await handleErrorsJson(response);
-      const tokenResponse = refreshTokenResponseSchema.safeParse(responseJson);
+      const tokenResponse = parseRefreshTokenResponse(responseJson, refreshTokenResponseSchema);
       o365AuthCredentials = { ...o365AuthCredentials, ...(tokenResponse.success && tokenResponse.data) };
-      if (!tokenResponse.success) {
-        console.error(
-          "Outlook error grabbing new tokens ~ zodError:",
-          tokenResponse.error,
-          "MS response:",
-          responseJson
-        );
-      }
       await prisma.credential.update({
         where: {
           id: credential.id,
diff --git a/packages/app-store/office365video/lib/VideoApiAdapter.ts b/packages/app-store/office365video/lib/VideoApiAdapter.ts
index 44c096b4a..06521d112 100644
--- a/packages/app-store/office365video/lib/VideoApiAdapter.ts
+++ b/packages/app-store/office365video/lib/VideoApiAdapter.ts
@@ -9,6 +9,7 @@ import type { PartialReference } from "@calcom/types/EventManager";
 import type { VideoApiAdapter, VideoCallData } from "@calcom/types/VideoApiAdapter";
 
 import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 
 let client_id = "";
 let client_secret = "";
@@ -57,16 +58,21 @@ const o365Auth = async (credential: CredentialPayload) => {
   const o365AuthCredentials = credential.key as unknown as O365AuthCredentials;
 
   const refreshAccessToken = async (refreshToken: string) => {
-    const response = await fetch("https://login.microsoftonline.com/common/oauth2/v2.0/token", {
-      method: "POST",
-      headers: { "Content-Type": "application/x-www-form-urlencoded" },
-      body: new URLSearchParams({
-        client_id,
-        refresh_token: refreshToken,
-        grant_type: "refresh_token",
-        client_secret,
-      }),
-    });
+    const response = await refreshOAuthTokens(
+      async () =>
+        await fetch("https://login.microsoftonline.com/common/oauth2/v2.0/token", {
+          method: "POST",
+          headers: { "Content-Type": "application/x-www-form-urlencoded" },
+          body: new URLSearchParams({
+            client_id,
+            refresh_token: refreshToken,
+            grant_type: "refresh_token",
+            client_secret,
+          }),
+        }),
+      "msteams",
+      credential.userId
+    );
 
     const responseBody = await handleErrorsJson<ITokenResponse>(response);
 
diff --git a/packages/app-store/salesforce/lib/CalendarService.ts b/packages/app-store/salesforce/lib/CalendarService.ts
index 4b16bb4c4..566d5dfb3 100644
--- a/packages/app-store/salesforce/lib/CalendarService.ts
+++ b/packages/app-store/salesforce/lib/CalendarService.ts
@@ -1,6 +1,7 @@
 import type { TokenResponse } from "jsforce";
 import jsforce from "jsforce";
 import { RRule } from "rrule";
+import { z } from "zod";
 
 import { getLocation } from "@calcom/lib/CalEventParser";
 import { WEBAPP_URL } from "@calcom/lib/constants";
@@ -16,6 +17,7 @@ import type {
 import type { CredentialPayload } from "@calcom/types/Credential";
 
 import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import parseRefreshTokenResponse from "../../_utils/oauth/parseRefreshTokenResponse";
 
 type ExtendedTokenResponse = TokenResponse & {
   instance_url: string;
@@ -34,6 +36,16 @@ const sfApiErrors = {
   INVALID_EVENTWHOIDS: "INVALID_FIELD: No such column 'EventWhoIds' on sobject of type Event",
 };
 
+const salesforceTokenSchema = z.object({
+  id: z.string(),
+  issued_at: z.string(),
+  instance_url: z.string(),
+  signature: z.string(),
+  access_token: z.string(),
+  scope: z.string(),
+  token_type: z.string(),
+});
+
 export default class SalesforceCalendarService implements Calendar {
   private integrationName = "";
   private conn: Promise<jsforce.Connection>;
@@ -60,6 +72,32 @@ export default class SalesforceCalendarService implements Calendar {
 
     const credentialKey = credential.key as unknown as ExtendedTokenResponse;
 
+    const response = await fetch("https://login.salesforce.com/services/oauth2/token", {
+      method: "POST",
+      body: new URLSearchParams({
+        grant_type: "refresh_token",
+        client_id: consumer_key,
+        client_secret: consumer_secret,
+        refresh_token: credentialKey.refresh_token,
+        format: "json",
+      }),
+    });
+
+    if (response.statusText !== "OK") throw new HttpError({ statusCode: 400, message: response.statusText });
+
+    const accessTokenJson = await response.json();
+
+    const accessTokenParsed = parseRefreshTokenResponse(accessTokenJson, salesforceTokenSchema);
+
+    if (!accessTokenParsed.success) {
+      return Promise.reject(new Error("Invalid refreshed tokens were returned"));
+    }
+
+    await prisma.credential.update({
+      where: { id: credential.id },
+      data: { key: { ...accessTokenParsed.data, refresh_token: credentialKey.refresh_token } },
+    });
+
     return new jsforce.Connection({
       clientId: consumer_key,
       clientSecret: consumer_secret,
diff --git a/packages/app-store/webex/lib/VideoApiAdapter.ts b/packages/app-store/webex/lib/VideoApiAdapter.ts
index f7f472911..37b0d3b19 100644
--- a/packages/app-store/webex/lib/VideoApiAdapter.ts
+++ b/packages/app-store/webex/lib/VideoApiAdapter.ts
@@ -8,6 +8,7 @@ import type { CredentialPayload } from "@calcom/types/Credential";
 import type { PartialReference } from "@calcom/types/EventManager";
 import type { VideoApiAdapter, VideoCallData } from "@calcom/types/VideoApiAdapter";
 
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import { getWebexAppKeys } from "./getWebexAppKeys";
 
 /** @link https://developer.webex.com/docs/meetings **/
@@ -58,18 +59,23 @@ const webexAuth = (credential: CredentialPayload) => {
   const refreshAccessToken = async (refreshToken: string) => {
     const { client_id, client_secret } = await getWebexAppKeys();
 
-    const response = await fetch("https://webexapis.com/v1/access_token", {
-      method: "POST",
-      headers: {
-        "Content-type": "application/x-www-form-urlencoded",
-      },
-      body: new URLSearchParams({
-        grant_type: "refresh_token",
-        client_id: client_id,
-        client_secret: client_secret,
-        refresh_token: refreshToken,
-      }),
-    });
+    const response = await refreshOAuthTokens(
+      async () =>
+        await fetch("https://webexapis.com/v1/access_token", {
+          method: "POST",
+          headers: {
+            "Content-type": "application/x-www-form-urlencoded",
+          },
+          body: new URLSearchParams({
+            grant_type: "refresh_token",
+            client_id: client_id,
+            client_secret: client_secret,
+            refresh_token: refreshToken,
+          }),
+        }),
+      "webex",
+      credential.userId
+    );
 
     const responseBody = await handleWebexResponse(response, credential.id);
 
diff --git a/packages/app-store/zoho-bigin/lib/CalendarService.ts b/packages/app-store/zoho-bigin/lib/CalendarService.ts
index 7f1ba1625..676c8062e 100644
--- a/packages/app-store/zoho-bigin/lib/CalendarService.ts
+++ b/packages/app-store/zoho-bigin/lib/CalendarService.ts
@@ -15,6 +15,7 @@ import type {
 import type { CredentialPayload } from "@calcom/types/Credential";
 
 import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import { appKeysSchema } from "../zod";
 
 export type BiginToken = {
@@ -81,11 +82,16 @@ export default class BiginCalendarService implements Calendar {
       refresh_token: credentialKey.refresh_token,
     };
 
-    const tokenInfo = await axios.post(accountsUrl, qs.stringify(formData), {
-      headers: {
-        "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
-      },
-    });
+    const tokenInfo = await refreshOAuthTokens(
+      async () =>
+        await axios.post(accountsUrl, qs.stringify(formData), {
+          headers: {
+            "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
+          },
+        }),
+      "zoho-bigin",
+      credentialId
+    );
 
     if (!tokenInfo.data.error) {
       // set expiry date as offset from current time.
diff --git a/packages/app-store/zohocrm/lib/CalendarService.ts b/packages/app-store/zohocrm/lib/CalendarService.ts
index e5a7e8fcf..ef5bb4ec8 100644
--- a/packages/app-store/zohocrm/lib/CalendarService.ts
+++ b/packages/app-store/zohocrm/lib/CalendarService.ts
@@ -16,6 +16,7 @@ import type {
 import type { CredentialPayload } from "@calcom/types/Credential";
 
 import getAppKeysFromSlug from "../../_utils/getAppKeysFromSlug";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 
 export type ZohoToken = {
   scope: string;
@@ -200,14 +201,19 @@ export default class ZohoCrmCalendarService implements Calendar {
           client_secret: this.client_secret,
           refresh_token: credentialKey.refresh_token,
         };
-        const zohoCrmTokenInfo = await axios({
-          method: "post",
-          url: url,
-          data: qs.stringify(formData),
-          headers: {
-            "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
-          },
-        });
+        const zohoCrmTokenInfo = await refreshOAuthTokens(
+          async () =>
+            await axios({
+              method: "post",
+              url: url,
+              data: qs.stringify(formData),
+              headers: {
+                "Content-Type": "application/x-www-form-urlencoded;charset=utf-8",
+              },
+            }),
+          "zohocrm",
+          credential.userId
+        );
         if (!zohoCrmTokenInfo.data.error) {
           // set expiry date as offset from current time.
           zohoCrmTokenInfo.data.expiryDate = Math.round(Date.now() + 60 * 60);
diff --git a/packages/app-store/zoomvideo/lib/VideoApiAdapter.ts b/packages/app-store/zoomvideo/lib/VideoApiAdapter.ts
index ef227596c..20e8f900e 100644
--- a/packages/app-store/zoomvideo/lib/VideoApiAdapter.ts
+++ b/packages/app-store/zoomvideo/lib/VideoApiAdapter.ts
@@ -9,6 +9,8 @@ import type { CredentialPayload } from "@calcom/types/Credential";
 import type { PartialReference } from "@calcom/types/EventManager";
 import type { VideoApiAdapter, VideoCallData } from "@calcom/types/VideoApiAdapter";
 
+import parseRefreshTokenResponse from "../../_utils/oauth/parseRefreshTokenResponse";
+import refreshOAuthTokens from "../../_utils/oauth/refreshOAuthTokens";
 import { getZoomAppKeys } from "./getZoomAppKeys";
 
 /** @link https://marketplace.zoom.us/docs/api-reference/zoom-api/meetings/meetingcreate */
@@ -74,17 +76,22 @@ const zoomAuth = (credential: CredentialPayload) => {
     const { client_id, client_secret } = await getZoomAppKeys();
     const authHeader = "Basic " + Buffer.from(client_id + ":" + client_secret).toString("base64");
 
-    const response = await fetch("https://zoom.us/oauth/token", {
-      method: "POST",
-      headers: {
-        Authorization: authHeader,
-        "Content-Type": "application/x-www-form-urlencoded",
-      },
-      body: new URLSearchParams({
-        refresh_token: refreshToken,
-        grant_type: "refresh_token",
-      }),
-    });
+    const response = await refreshOAuthTokens(
+      async () =>
+        await fetch("https://zoom.us/oauth/token", {
+          method: "POST",
+          headers: {
+            Authorization: authHeader,
+            "Content-Type": "application/x-www-form-urlencoded",
+          },
+          body: new URLSearchParams({
+            refresh_token: refreshToken,
+            grant_type: "refresh_token",
+          }),
+        }),
+      "zoomvideo",
+      credential.userId
+    );
 
     const responseBody = await handleZoomResponse(response, credential.id);
 
@@ -94,7 +101,7 @@ const zoomAuth = (credential: CredentialPayload) => {
       }
     }
     // We check the if the new credentials matches the expected response structure
-    const parsedToken = zoomRefreshedTokenSchema.safeParse(responseBody);
+    const parsedToken = parseRefreshTokenResponse(responseBody, zoomRefreshedTokenSchema);
 
     // TODO: If the new token is invalid, initiate the fallback sequence instead of throwing
     // Expanding on this we can use server-to-server app and create meeting from admin calcom account
